### PR TITLE
PRT-343 reach by default to maximum number of connections available 

### PR DESCRIPTION
**Pull request by ranlavanet**
* **Please check if the PR fulfills these requirements**
- [x] Commit message follows the Contribution Guidelines
- [x] Tests ran locally and added/modified if needed
- [x] Docs have been added/updated, if applicable
- [x] If applicable - JIRA ticket ID was added

* **What kind of change does this PR introduce?** (Bug fix, feature, unit tests, docs update, ...)
adding and removing clients as the provider gets heavier usage 

* **What is the current behavior?** (You can also link to an open issue here)
constant number of clients when starting

* **What is the new behavior (if this is a feature change)?**
adding and removing clients as the provider gets heavier usage 

* **Please describe what manual tests you ran, if applicable**
added unit tests and ran them 

* **Other information**:

### CODE REVIEW

File: relayer/chainproxy/connector.go [hunk 1]
```diff
@@ -101,16 +108,45 @@ func (connector *Connector) Close() {
...
 }
 
-func (connector *Connector) GetRpc(block bool) (*rpcclient.Client, error) {
+func (connector *Connector) increaseNumberOfClients(ctx context.Context) {
+	utils.LavaFormatDebug("No clients are free, increasing number of clients", nil)
+	var rpcClient *rpcclient.Client
+	var err error
+	for connectionAttempt := 0; connectionAttempt < MaximumNumberOfParallelConnectionsAttempts; connectionAttempt++ {
+		nctx, cancel := context.WithTimeout(ctx, DialTimeout)
+		rpcClient, err = rpcclient.DialContext(nctx, connector.addr)
```
Comment (omerlavanet) [inline relayer/chainproxy/connector.go:117]: new connector creating a client and this one have a lot of dup code, create an inner function and reuse it

File: relayer/chainproxy/connector.go [hunk 2]
```diff
@@ -101,16 +108,45 @@ func (connector *Connector) Close() {
 	}
 }
 
-func (connector *Connector) GetRpc(block bool) (*rpcclient.Client, error) {
+func (connector *Connector) increaseNumberOfClients(ctx context.Context) {
+	utils.LavaFormatDebug("No clients are free, increasing number of clients", nil)
```
Comment (omerlavanet) [inline relayer/chainproxy/connector.go:112]: this is misleading, we have still half free, how about we say how much?

File: relayer/chainproxy/connector.go [hunk 3]
```diff
@@ -178,16 +227,44 @@ func NewGRPCConnector(ctx context.Context, nConns uint, addr string) *GRPCConnec
...
+			cancel()
+			continue
+		}
+		cancel()
+
+		connector.lock.Lock() // add connection to free list.
+		defer connector.lock.Unlock()
+		connector.freeClients = append(connector.freeClients, grpcClient)
+		return
+	}
```
Comment (omerlavanet) [inline relayer/chainproxy/connector.go:248]: add a debug print in case we failed all retries?

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: relayer/chainproxy/connector.go (hunk 1 above)
Comment (ranlavanet) [reply relayer/chainproxy/connector.go:117]: not in this task.

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**